src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -21,180 +21,180 @@ internal sealed class TaskCenterSolutionAnalysisProgressReporter
-        private ProgressStatus _lastProgressStatus;
-        private ProgressStatus _lastShownProgressStatus;
-
-        // _resettableDelay makes sure that we at the end update status to correct value.
-        // in contrast to _lastTimeReported makes sure that we only update at least s_minimumInterval interval.
-        //
-        // for example, when an event stream comes in as below (assuming 200ms minimum interval)
-        // e1 -> (100ms)-> e2 -> (300ms)-> e3 -> (100ms) -> e4
-        //
-        // actual status shown to users without _resettableDelay will be 
-        // e1 -> e3.
-        //
-        // e2 and e4 will be skipped since interval was smaller than min interval.
-        // losing e2 is fine, but e4 is problematic since the user now could see the wrong status 
-        // until the next event comes in. 
-        // for example, it could show "Evaluating" when it is actually "Paused" until the next event
-        // which could be long time later.
-        // what _resettableDelay does is making sure that if the next event doesn't come in 
-        // within certain delay, it updates status to e4 (current).
-        private ResettableDelay _resettableDelay;
-
-        // this is only field that is shared between 2 events streams (IDiagnosticService and ISolutionCrawlerProgressReporter)
-        // and can be called concurrently.
-        private volatile ITaskHandler _taskHandler;
+        /// <summary>
+        /// Task handler to provide a task to the <see cref="_taskCenterService"/>
+        /// </summary>
+        private ITaskHandler? _taskHandler;
+
+        /// <summary>
+        /// Stores the currently running task center task.
+        /// This is manually started and completed based on receiving start / stop events
+        /// from the <see cref="ISolutionCrawlerProgressReporter"/>
+        /// </summary>
+        private TaskCompletionSource<VoidResult>? _taskCenterTask;
+
+        /// <summary>
+        /// Task used to trigger throttled UI updates in an interval
+        /// defined by <see cref="s_minimumInterval"/>
+        /// </summary>
+        private Task? _intervalTask;
+
+        /// <summary>
+        /// Gate access to reporting sln crawler events so we cannot
+        /// report UI changes concurrently.
+        /// </summary>
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Stores the last shown <see cref="ProgressData"/>
+        /// </summary>
+        private ProgressData _lastProgressData;
+
+        /// <summary>
+        /// Unfortunately, <see cref="ProgressData.PendingItemCount"/> is only reported
+        /// when the <see cref="ProgressData.Status"/> is <see cref="ProgressStatus.PendingItemCountUpdated"/>
+        /// So we have to store this in addition to <see cref="_lastProgressData"/> so that we
+        /// do not overwrite the last reported count with 0.
+        /// </summary>
+        private int _lastProgressCount;
```

Comment:
consider storing this and _lastProgresData as a tuple, so make it clear they update tgether.